python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll

```diff
@@ -25,3 +18,117 @@ predicate defaultAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nod
   or
   any(AdditionalTaintStep a).step(nodeFrom, nodeTo)
 }
+
+/**
+ * Holds if taint can flow in one local step from `nodeFrom` to `nodeTo` excluding
+ * local data flow steps. That is, `nodeFrom` and `nodeTo` are likely to represent
+ * different objects.
+ */
+predicate localAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+  concatStep(nodeFrom, nodeTo)
+  or
+  subscriptStep(nodeFrom, nodeTo)
+  or
+  stringMethods(nodeFrom, nodeTo)
+}
+
+/**
+ * Holds if taint can flow from `nodeFrom` to `nodeTo` with a step related to concatenation.
+ *
+ * Note that since we cannot easily distinguish interesting types (like string, list, tuple), so
+ * we consider any `+` operation to propagate taint. After consulting with the JS team, this
+ * should doesn't sound like it is a big problem in practice.
+ */
+predicate concatStep(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeTo) {
+  exists(BinaryExprNode add | add = nodeTo.getNode() |
+    add.getOp() instanceof Add and
+    (
+      add.getLeft() = nodeFrom.getNode()
+      or
+      add.getRight() = nodeFrom.getNode()
+    )
+  )
+}
+
+/**
+ * Holds if taint can flow from `nodeFrom` to `nodeTo` with a step related to subscripting.
+ */
+predicate subscriptStep(DataFlow::CfgNode nodeFrom, DataFlow::CfgNode nodeTo) {
+  nodeTo.getNode().(SubscriptNode).getObject() = nodeFrom.getNode()
+}
```

Comment:
No this is a general subscript step, which string slicing is a part of. How it plays together with content flow is still a bit unclear to me.

Reply:
I think content flow will handle moset cases for us (but not slicing, for instance), but I am not sure if it does exactly what we want. We should have a separate project to debate and test that. It seems we should leave the general step in here for now, since we know we want it for taint and it handles slicing. If at some point dataflow does the same approximation for all the relevant cases, we can consider taking it out...